storage2/src/state.rs

```diff
@@ -19,42 +19,58 @@ pub struct State {
     // A `None` value represents deletion.
     pub(crate) unwritten_changes: BTreeMap<String, Option<Vec<u8>>>,
     // A `None` value represents deletion.
-    pub(crate) sidecar_changes: BTreeMap<Vec<u8>, Option<Vec<u8>>>,
+    pub(crate) nonconsensus_changes: BTreeMap<Vec<u8>, Option<Vec<u8>>>,
 }
 
 impl State {
     pub(crate) fn new(snapshot: Arc<Snapshot>) -> Self {
         Self {
             snapshot,
             unwritten_changes: BTreeMap::new(),
-            sidecar_changes: BTreeMap::new(),
+            nonconsensus_changes: BTreeMap::new(),
         }
     }
 
     pub fn begin_transaction(&mut self) -> StateTransaction {
         StateTransaction::new(self)
     }
+
+    pub fn apply_transaction(&mut self, transaction: StateTransaction) -> Result<()> {
+        if transaction.failed {
+            return Err(anyhow::anyhow!("transaction failed").context(transaction.failure_reason));
+        }
+
+        // Write the unwritten consensus-critical changes to the state:
+        self.unwritten_changes.extend(transaction.unwritten_changes);
+
+        // Write the unwritten nonconsensus changes to the state:
+        self.nonconsensus_changes
+            .extend(transaction.nonconsensus_changes);
+
+        Ok(())
+    }
 }
 
 #[async_trait]
+// TODO: @hdevalence does this really need to be a trait?
```

Comment:
This was resolved in Discord, right? Might be worth putting a comment here about why: I actually forget why it needs to be a trait myself, again, but I know we reached consensus about it.